rust/src/pool/mod.rs

```diff
@@ -0,0 +1,409 @@
+use std::{
+    ops::DerefMut,
+    sync::{Arc, LockResult, RwLock},
+};
+
+use crate::error::Error;
+
+/// An abstraction for a pool of thread-safe elements
+pub trait Pool {
+    /// The key type used to identify entries in the pool.
+    type Key;
+    /// The type of elements indexed by the pool.
+    type Type;
+
+    /// Retrieves an entry from the pool.
+    /// The concrete type returned by the get operation may not be the concrete type `Self::Type`
+    /// but instead a wrapper type which behaves like `Self::Type`. This abstraction allows for
+    /// the pool to store complex types which track the lifetime of stored objects, enabling
+    /// nested catching.
+    fn get(
+        &self,
+        id: Self::Key,
+    ) -> Result<PoolEntry<impl DerefMut<Target = Self::Type> + Send + Sync + 'static>, Error>;
+
+    /// Stores the value in the pool and reserves an ID for it.
+    fn set(&self, value: Self::Type) -> Result<Self::Key, Error>;
+
+    /// Deletes the entry with the given ID from the pool
+    /// The ID may be reused in the future, when creating a new entry by calling [`Pool::set`].
+    fn delete(&self, id: Self::Key) -> Result<(), Error>;
+
+    /// Flushes all pending operations to the underlying IO hierarchy.
+    #[allow(dead_code)]
+    fn flush(&self) -> Result<(), Error>;
+}
+
+/// An element retrieved from the pool which can be locked before reading to enable safe concurrent
+/// access. Instances of this object shall not be kept for longer than the scope where they are
+/// needed, failing to release these nodes will lead to resources leackage.
+#[derive(Debug)]
+pub struct PoolEntry<T>(Arc<RwLock<T>>);
+
+impl<T> PoolEntry<T> {
+    /// Creates a new pool entry with the given [`NodePoolEntry`].
+    pub fn new(value: Arc<RwLock<T>>) -> Self {
+        Self(value)
+    }
+
+    /// Acquires a read lock on the entry.
+    pub fn read(&self) -> LockResult<std::sync::RwLockReadGuard<'_, T>> {
+        self.0.read()
+    }
+
+    /// Acquires a write lock on the entry.
+    pub fn write(&self) -> LockResult<std::sync::RwLockWriteGuard<'_, T>> {
+        self.0.write()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+
+    use std::{
+        collections::HashMap,
+        ops::{Deref, DerefMut},
+        sync::{Arc, Mutex, RwLock},
+        thread,
+    };
+
+    use crate::{
+        error::Error,
+        pool::{Pool, PoolEntry},
+        storage,
+    };
+
+    const TREE_DEPTH: u32 = 6;
+    const CHILDREN_PER_NODE: u32 = 3;
+
+    /// Recursively sets up a tree with a [`FakeNodePool`] with depth `MAX_DEPTH` and `MAX_CHILDREN`
+    /// children per node. Each node's value is set to a unique number based on its position in
+    /// the tree.
```

Comment:
The function operates on a pool, not a `FakeNodePool`, also the two constants no longer exist. And maybe it should mention that this is done using multiple threads - that's the main point, no?